src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,91 @@
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Class<? extends Annotation> annotationForInjection = Inject.class;
+    private final Class<? extends Annotation> annotationForImplementation = Component.class;
+    private final Map<Class<?>, Object> instances = new HashMap<>();
+    private final Map<Class<?>, Class<?>> interfaceImplementations = new HashMap<>();
+
+    {
+        interfaceImplementations.put(FileReaderService.class, FileReaderServiceImpl.class);
+        interfaceImplementations.put(ProductParser.class, ProductParserImpl.class);
+        interfaceImplementations.put(ProductService.class, ProductServiceImpl.class);
+    }
+
+    private Injector() {
+    }
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        compareClassAnnotationWithPattern(clazz, annotationForImplementation);
+        Field[] declaredFields = clazz.getDeclaredFields();
+
+        for (Field field : declaredFields) {
+            if (field.isAnnotationPresent(annotationForInjection)) {
+                Object fieldInstance = getInstance(field.getType());
+                clazzImplementationInstance = createNewInstance(clazz);
+                try {
+                    field.setAccessible(true);
+                    field.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialise field value. Class "
+                            + clazz.getName() + "Field: " + field.getName());
+                }
+            }
+        }
+        return clazzImplementationInstance == null ? createNewInstance(clazz)
+                : clazzImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(clazz, instance);
+            return instance;
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException("Can't create a new instance of " + clazz.getName(), e);
+        }
+    }
+
+    private Class<?> findImplementation(Class<?> interfaceClazz) {
+        if (interfaceClazz.isInterface()) {
+            if (!interfaceImplementations.containsKey(interfaceClazz)) {
+                throw new RuntimeException("Can't find implementation of "
+                        + interfaceClazz.getName() + "in map");
+            }
```

Comment:
Don't you use @Component andcompareClassAnnotationWithPattern()  for this pusrpose?

Reply:
> Don't you use @component andcompareClassAnnotationWithPattern() for this pusrpose?

In this case some tests don't pass
Tests expect returning exception from method getInstance()